akka-http-tests/src/test/scala/akka/http/scaladsl/marshallers/sprayjson/JsonStreamingSpec.scala

```diff
@@ -0,0 +1,170 @@
+    """.stripMargin
+
+  val arrayPretty =
+    """|[
+       |{"fname":"Frank","name":"Smith","age":42,"id":1337,"boardMember":false}
+       |{"fname":"Bob","name":"Smith","age":42,"id":1337,"boardMember":false}
+       |{"fname":"Hank","name":"Smith","age":42,"id":1337,"boardMember":false}
+       |]""".stripMargin
+
+  "JsonStreaming" should {
+    "read using entity(stream[T])" in {
+      val route = post {
+        entity(stream[Employee]) { employees ⇒
+          complete(employees.map(_.fname).intersperse(",").runFold("")(_ + _))
+        }
+      }
+
+      PostJson("/", lineByLineCommaSeparated) ~> route ~> check {
+        responseAs[String] shouldEqual "Frank,Bob,Hank"
+      }
+    }
+
+    "read entity(stream[JsObject])" in {
+      val route = post {
+        entity(stream[JsObject]) { js ⇒
+          complete(js.runFold("")({ case (acc, x) ⇒ acc + "." }))
+        }
+      }
+
+      PostJson("/", lineByLineCommaSeparated) ~> route ~> check {
+        status shouldEqual StatusCodes.OK
+      }
+    }
+
+    "reject when content type is not JSON" in {
+      val route = post {
+        entity(stream[Employee]) { employees ⇒
+          complete(employees.map(_.fname).intersperse(",").runFold("")(_ + _))
+        }
+      }
+
+      Post("/").withEntity(ContentTypes.`text/plain`, "Hello world!") ~> route ~> check {
+        handled shouldEqual false
+        rejection shouldEqual UnsupportedRequestContentTypeRejection(Set(ContentTypes.`application/json`))
+      }
+    }
+
+    "read using explicit framing" in {
+      val route = post {
+        entity(stream[Employee](framing = JsonFraming)) { employees ⇒
+          complete(employees.map(_.fname).intersperse(",").runFold("")(_ + _))
+        }
+      }
+
+      PostJson("/", lineByLineCommaSeparated) ~> route ~> check {
+        responseAs[String] shouldEqual "Frank,Bob,Hank"
+      }
+    }
+
+    "read in parallel (unordered) explicit framing" in {
+      val route = post {
+        entity(streamAsyncUnordered[Employee](parallelism = 3)) { employees ⇒
+          complete(employees.map(_.fname).intersperse(",").runFold("")(_ + _))
+        }
+      }
+
+      PostJson("/", lineByLineCommaSeparated) ~> route ~> check {
+        responseAs[String] shouldEqual "Frank,Bob,Hank"
+      }
+    }
+
+    "write streamed json line by line" in {
+      val frank = Employee.simple
+
+      val route = get {
+        complete(Source.repeat(frank).take(3))
+      }
+
+      val json =
+        """{"name":"Smith","boardMember":false,"fname":"Frank","age":42,"id":12345}""" + "\n" +
+          """{"name":"Smith","boardMember":false,"fname":"Frank","age":42,"id":12345}""" + "\n" +
+          """{"name":"Smith","boardMember":false,"fname":"Frank","age":42,"id":12345}""" // TODO what if someone wants eager newline?
+
+      Get("/") ~> route ~> check {
+        responseAs[String] shouldEqual json
+        contentType shouldEqual ContentTypes.`application/json`
```

Comment:
Hard to say. I haven't look into common approaches to JSON streaming so far.
But I'd say that something with content-type `application/json` should be legal JSON no matter what parameters are potentially attached to the content-type. That pretty much means that the top-level element must either be an array, whose elements are streamed (with commas in-between, as required) or an object, whose members are streamed (with commas in-between, as required). However, I'm not sure whether the latter approach is actually used anywhere in the wild.

If we break the JSON spec by introducing "funny" stream boundaries then we'd loose the ability to simply "re-chunk" the stream (e.g. by buffering). I think there is prettyness in *not* breaking the spec when streaming.